pkg/operator/controller/ingress/controller.go

```diff
@@ -713,6 +726,121 @@ func (r *reconciler) ensureIngressDeleted(ingress *operatorv1.IngressController)
+	return nil
+}
+
+// Clears this route's status that is admitted by provided ingressController
+func (r *reconciler) clearRouteStatus(route routev1.Route, icName string) (bool, error) {
+	// Go through each route and clear status if admitted by this ingress controller.
+	var updated routev1.Route
+	clearedStatus := false
+	for i := range route.Status.Ingress {
+		if condition := findCondition(&route.Status.Ingress[i], routev1.RouteAdmitted); condition != nil {
+			if routerName := &route.Status.Ingress[i].RouterName; routerName != nil && *routerName == icName {
+				// Remove this status since it matches our routerName.
+				route.DeepCopyInto(&updated)
+				updated.Status.Ingress = append(route.Status.Ingress[:i], route.Status.Ingress[i+1:]...)
+				if err := r.client.Status().Update(context.TODO(), &updated); err != nil {
+					return clearedStatus, fmt.Errorf("failed to clear route status for %s with routerName %s: %v", route.Name, icName, err)
+				}
+				clearedStatus = true
+			}
+		}
+	}
+
+	return clearedStatus, nil
+}
+
+// Clear routes that are not owned by this ingress controller
+func (r *reconciler) clearRoutesNotAdmittedByIngress(ingress *operatorv1.IngressController) {
+	// TODO: THIS IS TESTING CODE. NEEDS A LOT OF WORK
+	start := time.Now()
+
+	routesCleared := 0
+	// List all routes.
+	routeList := &routev1.RouteList{}
+	if err := r.cache.List(context.TODO(), routeList); err != nil {
+		fmt.Errorf("failed to list all routes in order to clear route status")
+	}
+	for _, route := range routeList.Items {
+		routeSelectorMatch := false
+		namespaceSelectorMatch := false
+
+		// TODO: REAL MESSY, FIX THIS UP
+		routeSelector, _ := metav1.LabelSelectorAsSelector(ingress.Spec.RouteSelector)
+		if routeSelector != nil {
+			routeLabelSelector, _ := labels.Parse(routeSelector.String())
+			if routeLabelSelector == nil {
+				routeSelectorMatch = true
+			} else if routeLabelSelector.Matches(labels.Set(route.Labels)) {
+				routeSelectorMatch = true
+			}
+		} else {
+			routeSelectorMatch = true
+		}
+
+		// TODO: What if namespace Expression is invalid? This will panic
+		namespaceSelector, _ := metav1.LabelSelectorAsSelector(ingress.Spec.NamespaceSelector)
+
+		namespaceLabelSelector, _ := labels.Parse(namespaceSelector.String())
+		ns := &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: route.Namespace}}
+		if err := r.cache.Get(context.TODO(), types.NamespacedName{Name: route.Namespace}, ns); err != nil {
+			fmt.Errorf("failed to get namespace of route")
+		}
+		if namespaceLabelSelector == nil {
+			namespaceSelectorMatch = true
+		} else if namespaceLabelSelector.Matches(labels.Set(ns.Labels)) {
+			namespaceSelectorMatch = true
+		}
+		// If not my route, then clear my state (if any...)
+		if !namespaceSelectorMatch || !routeSelectorMatch {
+			fmt.Printf("clearing route %s from ingress controller %s", route.Name, ingress.Name)
+			cleared, err := r.clearRouteStatus(route, ingress.ObjectMeta.Name)
```

Comment:
I'm tempted to suggest you just clear the route status unconditionally and force the router to re-admit all the routes, but that would probably cause too much churn.

Reply:
Agreed. I suppose it wouldn't be too bad when we update it to only clear the status when the selectors are updated (that should be very rare right?). It would majorly reduce the logic here, that would be a benefit. But would not be the most efficient solution.